### ISSUE TITLE
Comparison, tunning, and Cross Validation

### ISSUE THREAD

**Issue by Jason-cse** [Jason-cse/ML-wine_quality#19]
`tuning_spaces = {
    'Random Forest': (RandomForestRegressor(random_state=RANDOM_STATE, n_jobs=-1), {'model__n_estimators': [200, 400], 'model__max_depth': [None, 12, 24], 'model__min_samples_leaf': [1, 2]}),
    'Gradient Boosting': (GradientBoostingRegressor(random_state=RANDOM_STATE), {'model__n_estimators': [100, 200], 'model__learning_rate': [0.03, 0.08], 'model__max_depth': [2, 3]}),
    'Ridge': (Ridge(), {'model__alpha': [0.01, 0.1, 1, 10, 100]})
}
tuned_results = []
for name, (model, params) in tuning_spaces.items():
    search = GridSearchCV(make_pipeline(model), params, scoring='neg_root_mean_squared_error', cv=cv, n_jobs=-1)
    search.fit(X_train, y_train)
    predictions = search.predict(X_test)
    tuned_results.append({'model': name, 'best CV RMSE': -search.best_score_, 'test RMSE': mean_squared_error(y_test, predictions) ** 0.5, 'test MAE': mean_absolute_error(y_test, predictions), 'test R2': r2_score(y_test, predictions), 'best params': search.best_params_})
display(pd.DataFrame(tuned_results).sort_values('test RMSE'))`
